### ISSUE TITLE
horovod with tensorflow, PBS, MPI and an ssh identity file fails cause of shuffled arguments for ssh

### ISSUE THREAD

**Issue by afluegel9** [horovod/horovod#3823]
**Environment:**
1. Framework: tensorflow
2. Framework version: 2.8.2
3. Horovod version: 0.26.1
4. MPI version: openmpi-4.1.2_SLE15_ucx-1.C.1
5. CUDA version: 11.7
6. NCCL version: not used
7. Python version: 3.9.6
8. Spark / PySpark version: not used
9. Ray version: not used
10. OS and version: Linux SLES 15 SP 3
11. GCC version: 7.5.0
12. CMake version: 3.17.0

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)? no hang.
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? no docker used
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? sure

**Bug report:**
We use horovodrun to submit to PBS with mpirun and openMPI.
The PBS submit is (reduced to the beef):
qsub -j n -m eb -o stdout.$$ -e stderr.$$ -N HOROVOD -P my-project -l longname=HOROVOD -q our-cluster -l ncpus=40 -l mem=100gb -l place=scatter:excl << "EOF"
cd "$PBS_O_WORKDIR"
# construct argument for -H of horovodrun
HOSTS_FOR_HOROVOD=""
SEP=""
for H in `cat "$PBS_NODEFILE"` ; do
  HOSTS_FOR_HOROVOD="$HOSTS_FOR_HOROVOD$SEP$H:1"
  SEP=","
done

# environment for PBS and open MPI
unset PBS_ENVIRONMENT
unset PBS_NODEFILE

export OMPI_MCA_plm=rsh
export OMPI_MCA_plm_rsh_agent=$PBS_EXEC/bin/pbs_remsh  # rsh command used for openmpi
export PBS_RSHCOMMAND="/usr/bin/ssh -i $HOME/.ssh/id_mpi -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o ChallengeResponseAuthentication=no"                               # rsh command for pbs_attach called from pbs_remsh

exec horovodrun -i $HOME/.ssh/id_mpi --mpi --mpi-threads-disable -np 5 -H $HOSTS_FOR_HOROVOD python tf_cnn_benchmarks.py --data_format=NCHW --device=CPU --xla=false --num_intra_threads=16 --num_inter_threads=2 --mkl=true --variable_update=horovod --data_dir=$HOME/tensorflow/benchmark/data/cifar-10-batches-py --data_name=cifar10 --model=resnet110 --distortions=true --optimizer=sgd -num_warmup_batches=1 --display_every=10 --print_training_accuracy=true --num_epochs=1 --batch_size=1024 --init_learning_rate=3.20 --learning_rate_decay_factor=0.9 --num_epochs_per_decay=5 --minimum_learning_rate=0.01 --num_learning_rate_warmup_epochs=1
EOF

In the stderr.$$ file i see the following error for each of the hosts in $PBS_NODEFILE:
Warning: Identity file PBS_CONF_FILE=/etc/pbs.conf not accessible: No such file or directory.
unknown option -- j
usage: ssh [-46AaCfGgKkMNnqsTtVvXxYy] [-B bind_interface]
           [-b bind_address] [-c cipher_spec] [-D [bind_address:]port]
           [-E log_file] [-e escape_char] [-F configfile] [-I pkcs11]
           [-i identity_file] [-J [user@]host[:port]] [-L address]
           [-l login_name] [-m mac_spec] [-O ctl_cmd] [-o option] [-p port]
           [-Q query_option] [-R address] [-S ctl_path] [-W host:port]
           [-w local_tun[:remote_tun]] destination [command]

Huuuh ? Where's the -j coming from ? Why is PBS_CONF_FILE=/etc/pbs.conf considered the identity file ?
stracing shows this call to ssh (anonymized):
[pid 3037] execve("/usr/bin/ssh", ["/usr/bin/ssh", "-i", "/home/myusername/.ssh/id_mpi", "-o", "StrictHostKeyChecking=no", "-o", "KbdInteractiveAuthentication=no", "-o", "ChallengeResponseAuthentication=no", "-n", "-i", "PBS_CONF_FILE=/etc/pbs.conf", "/opt/pbs/bin/pbs_attach", "-j", "1307246.node289", "/home/myusername/.ssh/id_mpi", "node1", "orted", "-mca", "ess", "\"env\"", "-mca", "ess_base_jobid", "\"2484797440\"", "-mca", "ess_base_vpid", "3", "-mca", "ess_base_num_procs", "\"5\"", "-mca", "orte_node_regex", "\"node[4:1,2,3,4]@0(5)\"", "-mca", "orte_hnp_uri", "\"2484797440.0;tcp://10.10.10.10:54331\"", "-mca", "pml", "\"ob1\"", "-mca", "btl", "\"^openib\"", "-mca", "btl_tcp_if_include", "\"bond0\"", "-mca", "plm", "\"rsh\"", "--tree-spawn", "-mca", "routed", "\"radix\"", "-mca", "orte_parent_uri", "\"2484797440.0;tcp://10.10.10.10:54331\"", "-mca", "plm_rsh_agent", "\"/opt/pbs/bin/pbs_remsh\"", "-mca", "plm_rsh_args", "\"-i", "/home/myusername/.ssh/id_mpi\"", "-mca", "orte_tag_output", "\"1\"", "-mca", "hwloc_base_binding_policy", "\"none\"", "-mca", "rmaps_base_mapping_policy", "\"slot\"", "-mca", "pmix", "\"^s1,s2,cray,isolated\""], ["KMP_VERSION=TRUE", "OS=Linux_x86_64", "LD_LIBRARY_PATH=/opt/openmpi/openmpi-4.1.2_SLE15_ucx-1.C.1/lib", "HOSTTYPE=x86_64", "PBS_O_LANG=en_US.UTF-8", "PBS_SERVER=node1", "LANG=en_US.UTF-8", "HFI_NO_BACKTRACE=1", "PBS_O_SLOTS_PER_VNODE=8", "HOSTNAME=node1", "OLDPWD=/home/myusername/proj/tensorflow-tests/test-programs/cifar10/cpu", "PBS_O_HOME=/home/myusername", "AWS_PROFILE=MYUSERNAME", "CSHEDIT=emacs", "PBS_JOBID=1307246.node289", "ENVIRONMENT=BATCH", "GPG_TTY=not a tty", "HOROVOD_MPI_THREADS_DISABLE=1", "COLORTERM=1", "PBS_JOBNAME=HOROVOD", "NCPUS=8", "OMPI_MCA_orte_tag_output=1", "PBS_EXEC=/opt/pbs", "MACHTYPE=x86_64-suse-linux", "PBS_O_PATH=/bin:/usr/bin:/usr/sbin:/usr/lib/mit/bin:"..., "HN=node289", "OSTYPE=linux", "OMPI_MCA_plm=rsh", "PBS_O_WORKDIR=/home/myusername/proj/tensorflow-tests/test-programs/cifar10/cpu", "USER=myusername", "PAGER=less", "LC_COLLATE=C", "OMPI_MCA_plm_rsh_agent=/opt/pbs/bin/pbs_remsh", "MORE=-sl", "PBS_TASKNUM=1", "PWD=/home/myusername/proj/tensorflow-tests/tensorflow/benchmark/benchmarks-master/scripts/tf_cnn_benchmarks", "OMPI_MCA_plm_rsh_args=-i /home/myusername/.ssh/id_mpi", "HOME=/home/myusername", "HOST=node289.ourdomain.org", "XNLSPATH=/usr/share/X11/nls", "PBS_MOMPORT=10003", "IPATH_NO_BACKTRACE=1", "HISTFILE=/home/myusername/.sh_history_linux", "MKL_DEBUG=0", "PBS_JOBCOOKIE=0000000041D42137000000004ADE03A9", "PBS_O_SHELL=/bin/bash", "PROFILEREAD=true", "TMPDIR=/share/tmp/scratch", "PBS_HOME=/opt/pbs", "PMIX_MCA_mca_base_component_show_load_errors=1", "OMPI_MCA_pml=ob1","MAIL=/home/myusername/.mbox", "LESSKEY=/etc/lesskey.bin", "SHELL=/bin/bash", "OMPI_MCA_btl_tcp_if_include=bond0", "MKL_VERBOSE=1", "OMPI_MCA_hwloc_base_binding_policy=none", "OMPI_MCA_rmaps_base_mapping_policy=slot", "PYTHONSTARTUP=/etc/pythonstart", "SHLVL=3", "PBS_O_HOST=node289.ourdomain.org", "G_FILENAME_ENCODING=@locale,UTF-8,ISO-8859-15,CP1252", "MKLDNN_VERBOSE=0", "PBS_O_SYSTEM=Linux", "MANPATH=/usr/share/man:/opt/util/SLE15/usr/share/man", "PBS_O_LOGNAME=myusername", "PBS_NODENUM=0", "PBS_JOBDIR=/home/myusername", "LOGNAME=myusername", "PATH=/home/myusername/tf282_cpu/bin:/bin:/usr/bin:/usr/sbin:/usr/lib/"..., , "PBS_QUEUE=our-cluster", "G_BROKEN_FILENAMES=1", "HISTSIZE=1000", "CPU=x86_64", "PBS_RSHCOMMAND=/usr/bin/ssh -i /home/myusername/.ssh/id_mpi -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o ChallengeResponseAuthentication=no", "PBS_O_MAIL=/home/myusername/.mbox", "OMP_NUM_THREADS=1", "LESSOPEN=lessopen.sh %s", "OMPI_MCA_btl=^openib" ] <unfinished ...>

I would assume, that the order of the first arguments is wrong, one "-i" and one time the id-file is superfluous and all in all it should be:
"/usr/bin/ssh", "-i", "/home/myusername/.ssh/id_mpi", "-o", "StrictHostKeyChecking=no", "-o", "KbdInteractiveAuthentication=no", "-o", "ChallengeResponseAuthentication=no", "-n", "node1", "PBS_CONF_FILE=/etc/pbs.conf", "/opt/pbs/bin/pbs_attach", "-j", "1307246.node289", "orted", "-mca", ...

The issue might be in https://github.com/horovod/horovod/blob/master/horovod/runner/util/remote.py , but i'm not sure.
However, the problem does not occur, if no identity file needs to be given and a default identity file can be used or in general password-less ssh is possible without supplying an identity file explicitely on the commandline.
The issue can be avoided by inserting one line to [horovod/runner/launch.py](https://github.com/horovod/horovod/blob/master/horovod/runner/launch.py#L85):
*** /home/myusername/tf282_gpu/lib/python3.9/site-packages/horovod/runner/launch.py   2023-01-19 14:33:04.350538000 +0100
--- /home/myusername/tf282_cpu/lib/python3.9/site-packages/horovod/runner/launch.py   2023-01-24 18:09:51.209094000 +0100
***************
*** 84,87 ****
--- 84,88 ----
          return exit_code, output_msg
  
+     return True
      args_list = [[get_remote_command(local_command='true',
                                       host=host_address,

This just skips the ssh test.

**Comment by afluegel9** [comment]
In the meantime i found out, that the problem seems to be within mpirun of openmpi. It passes the arguments incorrectly, when the environment variable OMPI_MCA_plm_rsh_agent is set to $PBS_EXEC/bin/pbs_remsh . In this constellation (probably also in others) still ssh is invoked by mpirun with the shown mixture of arguments.
